### ISSUE TITLE
Cannot run multiple unprivileged systemd-nspawn containers simultaneously

### ISSUE THREAD

**Issue by AlexMekkering** [systemd/systemd#8599]
### Submission type

 - Bug report

### systemd version the issue has been seen with

> systemd 238
+PAM -AUDIT -SELINUX -IMA -APPARMOR +SMACK -SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +SECCOMP +BLKID +ELFUTILS +KMOD -IDN2 +IDN +PCRE2 default-hierarchy=hybrid

> Actually, systemd 238 commit 738ab7502afb7663d9aacdd73e79025aa7cd0a9b works just fine while the current branch [238-stable](https://github.com/systemd/systemd-stable/commits/v238-stable) (most recent commit [c58ab03](https://github.com/systemd/systemd-stable/commit/c58ab03f64890e7db88745a843bd4520e307099b)) from [systemd/systemd-stable](https://github.com/systemd/systemd-stable) has this problem.

### Used distribution

> Arch Linux (with package systemd version 238.51-1 installed which is based on [238-stable](https://github.com/systemd/systemd-stable/commits/v238-stable), commit [c58ab03](https://github.com/systemd/systemd-stable/commit/c58ab03f64890e7db88745a843bd4520e307099b))

### In case of bug report: Expected behaviour you didn't see

> Multiple unprivileged systemd-nspawn containers can be running simultaneously

### In case of bug report: Unexpected behaviour you saw

> After an unprivileged systemd-nspawn container (e.g. container1) has been started, no subsequent container (e.g. container2) can be started and the journal shows:
mrt 28 08:42:56 localhost systemd[1]: Starting Container container2...
-- Subject: Unit systemd-nspawn@container2.service has begun start-up
-- Defined-By: systemd
-- Support: https://lists.freedesktop.org/mailman/listinfo/systemd-devel
-- 
-- Unit systemd-nspawn@container2.service has begun starting up.
mrt 28 08:42:56 localhost systemd-nspawn[11295]: Selected user namespace base 201129984 and range 65536.
mrt 28 08:42:56 localhost systemd-nspawn[11295]: Failed to add new veth interfaces (ve-container2:host0): File exists
mrt 28 08:42:56 localhost systemd[1]: systemd-nspawn@container2.service: Main process exited, code=exited, status=1/FAILURE
mrt 28 08:42:56 localhost systemd[1]: systemd-nspawn@container2.service: Failed with result 'exit-code'.
mrt 28 08:42:56 localhost systemd[1]: Failed to start Container container2.
-- Subject: Unit systemd-nspawn@container2.service has failed
-- Defined-By: systemd
-- Support: https://lists.freedesktop.org/mailman/listinfo/systemd-devel
-- 
-- Unit systemd-nspawn@container2.service has failed.
-- 
-- The result is RESULT.

After container1 has been started, an interface named `host0` is listed by networkctl (executed on the host, so not within the container), which may be why it can't create it again:
> [root@nuc machines]# networkctl
IDX LINK             TYPE               OPERATIONAL SETUP     
  1 lo               loopback           carrier     unmanaged 
  2 eno1             ether              routable    configured
  3 host0            ether              off         unmanaged 
  4 ve-container1    ether              no-carrier  configuring

Also, when container1 is stopped, interface `host0` disappears and container2 can be started just fine (but again also results in a listed `host0` interface). In that case container1 can't be started.

Could this be a side-effect of #8430 ?

### In case of bug report: Steps to reproduce the problem

> On Arch Linux as root:
```
chdir /var/lib/machines
# Create containers
mkdir container1
pacstrap -c container1 systemd
mkdir container2
pacstrap -c container2 systemd

# Start container1 unprivileged (by default)
machinectl start container1
# Observe a listed host0 interface
networkctl

# Try to start container2
machinectl start container2
# Observe the fasilure
```

**Comment by Toolybird** [comment]
Similar issue for me. Except my containers are --private-users=no.

I'm also thinking #8430 looks suspicious.

Initial Arch forum report:

https://bbs.archlinux.org/viewtopic.php?id=235740

**Comment by qmega** [comment]
I have what I think is the same issue. Even with only one container, using `--network-veth` I get no network connectivity from inside the container. Like the report here, I see `host0` *outside* the container (and ignored by networkd), when I believe it's supposed to be *inside*. That seems like it'd be the root of the problem with starting multiple containers too.

I can see #8430 causing this, if it makes the unshare happen after the creation of `host0` where it used to happen before. I'd like to provide more than speculation, but I've downgraded for the moment as I need this for work. Happy to test patches / bisect this weekend if it's not resolved by then.

**Comment by SjonHortensius** [comment]
This fails for me using either veth or macvlan interfaces. The first container starts but has no network, any following containers fail with either `Failed to add new veth interfaces (vb-xxx:host0): File exists` or `Failed to add new macvlan interfaces: Device or resource busy`

**Comment by qmega** [comment]
Confirmed that reverting 0441378080489e4ab6704cd0a2d78cb1ceaca899 fixes this for me. Keeping the `unshare(CLONE_NEWNET)` separate but moving it before the writing of the UID map also works, but then /proc/net/ip_tables_names belongs to nobody so it's still effectively reverting the intended change of #8430.

Okay, actually, I think I figured this out. There's a race, and `host0` is "moved" into the container pid's network namespace before it gets unshared, so it stays in the host namespace. I got it to work by adding another barrier between the unshare in the child and the private network setup in the parent. (And ip_tables_names is still owned by root.) I'm not sure if that's the best way to fix this, but I'll try and clean it up for a PR in case it is. Hopefully it'll at least save some time for someone more familiar with the code.
